src/main.js

```diff
@@ -1625,6 +1624,41 @@ function discoveryFeedCardsFromQueries(proxies) {
     }).catch(e => logError(e, 'Query failed'));
 }
 
+
+// A helper class for tracking single-connections to signals
+//
+// A bit gross in the sense that it uses O(N) lookups, but
+// JS doesn't seem to have unique IDs for objects.
+const SignalManager = new Lang.Class({
+    Name: 'SignalManager',
+
+    _init: function() {
+        this._tracking = [];
+    },
+
+    track: function(obj, signal, callback) {
+        let conn = obj.connect(signal, callback);
+        this._tracking.push([signal, obj, conn]);
+        return conn;
+    },
+
+    drop: function(obj, signal) {
+        let pendingDisconnections = [];
+
+        this._tracking = this._tracking.filter(([trackedSignal, trackedObj, conn]) => {
+            if (trackedSignal === signal && trackedObj === obj) {
+                pendingDisconnections.push([trackedObj, conn]);
+                return false;
+            }
+
+            return true;
+        });
+
+        pendingDisconnections.forEach(([trackedObj, conn]) => trackedObj.disconnect(conn));
+    }
+});
```

Comment:
per @cosimoc 
> I think this class is more confusing than anything to be honest... It's a pretty well-established practice to store signal connection IDs in the main class like that, and this adds cognitive overhead for little gain IMO. I'd rather keep it simple.

per @smspillaz

> Yeah ... hmmm. I appreciate that the practice exists but I really don't like it ;-) I kind of see it as boilerplate.
> I wonder if there's a better way to do this rather than explicitly checking every single signal to be == -1 every time we want to reconnect it?

per @cosimoc 

> The practice is only really needed when the object emitting signals outlives the signal receiver, which is not the case most of the time. In C it's possible to use g_signal_connect_object() to make things simpler; not sure if that works from JS. Either way, this is also a different kind of boilerplate ;-) I'm not arguing against having a better convenience for that use case in JS if possible, but I wouldn't want to special-case it here.

Reply:
Indeed, it isn't possible to use `g_signal_connect_object` here.

What I really want to avoid (and in all of our long-running JS code specifically) is having to repeat the following block all the time:

    object._some_signal_id = -1;
    
    if (object._some_signal_id !== -1) {
        source.disconnect(object._some_signal_id);
        object._some_signal_id = -1;
     }

Its fine for one or two signals but it gets really repetitive for N signals. It worse if you have to do it in multiple places. Not to mention the fact that it is error prone - you have to explicitly check if the signal is unset, then disconnect it using the right source object, then explicitly set it to -1. What I really want to do is to be able to tell the object "if you were listening for this signal, do whatever you need to do to stop doing that now".

Obviously, if there are other good bits of this PR that we should include, then this commit shouldn't block that and we should separate it out for further discussion. But longer term I would like to find a way to do this that isn't the way that we're doing it right now, since the way that we're doing it right now strikes me as highly unergonomic and error prone.

Reply:
Okay - I have moved the signal manager stuff to a separate branch.